### PR TITLE
JSON with releases

### PR DESCRIPTION
**Pull request by xprazak2**
JSON file with info about released versions gets generated by Jekyll. Its content is based on already existing docs but we can add whatever we like to the misc.yaml. For now, we have core versions, latest releases of core and versions of plugins that have docs on wiki.

### CODE REVIEW

File: _plugins/versions_generator.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
+require 'json'
+require 'yaml'
+
+module Jekyll
+  class VersionsGenerator < Generator
```
Comment (domcleal) [inline _plugins/versions_generator.rb]: Custom plugins can't be run on GitHub Pages, I don't think it's a good idea to add one.

**Comment by xprazak2** [comment]
I removed the generator, now we have only static files.

File: _data/misc.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+releases:
```
Comment (domcleal) [inline _data/misc.yaml]: This file should probably be called releases, since that's what it contains.

File: _data/misc.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  core:
+    latest:
+      1-11: 1.11.0-RC1
+      1-10: 1.10.2
+      1-9: 1.9.3
+    all:
```
Comment (domcleal) [inline _data/misc.yaml]: What's this list for?  Should it contain pre-releases?

File: _data/misc.yaml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      - "1.3"
+      - "1.4"
+      - "1.5"
+      - "1.6"
+      - "1.7"
+      - "1.8"
+      - "1.9"
+      - "1.10"
+      - "1.11"
+  plugins:
```
Comment (domcleal) [inline _data/misc.yaml]: This list's incomplete, is it only going to contain plugins with documentation in this repo?

File: _data/misc.yaml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    foreman_ansible:
+      - "0.x"
```
Comment (domcleal) [inline _data/misc.yaml]: What's the format of these fields?  It seems limited to only major and/or minor versions, but the Foreman releases above contains patch revisions.

I'm not sure how it'd be maintained if patch revisions _were_ added though.

Comment (lzap) [inline _data/misc.yaml]: +1 Discovery does a lot of minor revisions, for example the first official release for 1.11 RC will be 5.0.1. We should be able to list them all there. Hopefully Foreman shows the incoming plugin updates to users too.

Comment (lzap) [inline _data/misc.yaml]: By the way you can get the lists from https://rubygems.org/gems/foreman_discovery/versions

File: _includes/version.html [hunk 6]
```diff
@@ -1,8 +1,8 @@
 <div class="version-marker">
   <p>
     Latest versions:
-      <a href="{{ site.baseurl }}documentation.html#1.11">1.11.0-RC1</a>
+      <a href="{{ site.baseurl }}documentation.html#1.11">{{ site.data.misc.releases.latest.1-11 }}</a>
```
Comment (lzap) [inline _includes/version.html]: Aren't you missing `core` in the variable name?

File: _data/misc.yaml (hunk 5 above)
Comment (domcleal) [inline _data/misc.yaml]: The other complication with plugins is that not all updates are applicable to all versions.  The "latest" plugin release might only be available in a nightly or later distribution, but the user would have to update Foreman before it makes sense to install it.

**Comment by lzap** [comment]
In manual and plugin layouts we use some tricks to show warning boxes for old or unreleased documentation. Since you introduce now this data, this could be used to actually show these boxes as well if you provide "latest" versions for all the plugins.

https://github.com/theforeman/theforeman.org/blob/gh-pages/_layouts/plugin.html#L11-Lundefined

File: _data/misc.yaml (hunk 5 above)
Comment (lzap) [inline _data/misc.yaml]: Yeah, maybe we could implement plugin matrix for each individual item and then render those for each plugin. Discovery is a good example of maintaining this in a table if you want some input data. Then we could use the data also to provide correct notifications to users about incoming plugin updates.

**Comment by xprazak2** [comment]
I added the missing versions of core into the list. I think the current format will not be too difficult to maintain, even with patch versions. I do not see a reason why it should not contain pre-releases, since we currently advertise 1.11.0-RC1 as latest version on the front page.

I also plugin into the list together with required foreman version. It should allow us to determine the latest plugin version for the foreman version that user has.

File: _data/releases.yml [hunk 7]
```diff
@@ -0,0 +1,674 @@
...
+    - 0.0.2rc1
+    - 0.0.1
+    - 0.0.1-6
+    - 0.0.1-5
+    - 0.0.1-4
+    - 0.0.1-3
+    - 0.0.1-2
+    - 0.0.1-1
+plugins:
+  foreman-digitalocean:
```
Comment (domcleal) [inline _data/releases.yml]: Why hyphenated?  The engine and gem name are both lower case.

File: _data/releases.yml [hunk 8]
```diff
@@ -0,0 +1,674 @@
...
+  foreman-docker:
+    1.4.0:
+      requires_foreman: '1.4'
+    0.0.3:
+      requires_foreman: '1.4'
+    0.0.2:
+      requires_foreman: '1.4'
+    0.0.1:
+      requires_foreman: '1.4'
+    2.0.0:
```
Comment (domcleal) [inline _data/releases.yml]: Is this list meant to be sorted?  It'll be easier to maintain.

File: _data/releases.yml [hunk 9]
```diff
@@ -0,0 +1,674 @@
...
+      requires_foreman: '1.4'
+    1.0.0:
+      requires_foreman: '1.4'
+    0.2.0:
+      requires_foreman: '1.4'
+  foreman-one:
+    0.0.4:
+      requires_foreman: '1.8'
+  foreman-tasks:
+    rubygem-foreman-tasks-0.6.1-1:
```
Comment (domcleal) [inline _data/releases.yml]: Looks incorrect...

File: _data/releases.yml [hunk 10]
```diff
@@ -0,0 +1,674 @@
...
+    1.0.1:
+      requires_foreman: '1.4'
+    1.0.0:
+      requires_foreman: '1.4'
+    0.2.0:
+      requires_foreman: '1.4'
+  foreman-one:
+    0.0.4:
+      requires_foreman: '1.8'
+  foreman-tasks:
```
Comment (domcleal) [inline _data/releases.yml:121]: also, gem or engine name?

**Comment by xprazak2** [comment]
Ah, inconsistencies in repo names, engine names and gems and in some cases even releases. Great. Will be back with changes

**Comment by domcleal** [comment]
Yeah, it's not great, but it happens despite the advice.  I would tend towards using the name that is registered in Foreman's plugin database at runtime.

**Comment by xprazak2** [comment]
foreman-taks is the only plugin that has hyphen in name, which can be seen at runtime, so I kept it here as well. Everything else should have underscore now and be sorted in descending order.

File: _data/releases.yml [hunk 11]
```diff
@@ -0,0 +1,822 @@
...
+  - 0.0.2rc1
+  - 0.0.1
+  - 0.0.1-6
+  - 0.0.1-5
+  - 0.0.1-4
+  - 0.0.1-3
+  - 0.0.1-2
+  - 0.0.1-1
+:plugins:
+  foreman_digitalocean:
```
Comment (domcleal) [inline _data/releases.yml]: File not alphabetised?

File: _data/releases.yml [hunk 12]
```diff
@@ -0,0 +1,822 @@
+---
+1:
+  11:
+  - 1.11.0-RC1
```
Comment (domcleal) [inline _data/releases.yml]: missing RC2

**Comment by domcleal** [comment]
The branch doesn't merge cleanly, please rebase it.

> foreman-taks is the only plugin that has hyphen in name, which can be seen at runtime

So going back to my earlier questions, which name is being used for this data?  The gem name?  (Since foreman-tasks' engine name is `foreman_tasks`, with an underscore.)

**Comment by xprazak2** [comment]
The gem name is used. It is also the one that is registered in Foreman runtime database afaik.

**Comment by xprazak2** [comment]
This feature will be replaced by [rss feed dashboard widget](https://github.com/theforeman/foreman/pull/3400)

**Pull request closed without merging**